addons/mrp/models/mrp_bom.py

```diff
@@ -157,7 +190,13 @@ def check_cycle(v, visited, recStack, graph):
                 continue
 
             line_quantity = current_qty * current_line.product_qty
-            bom = self._bom_find(product=current_line.product_id, picking_type=picking_type or self.picking_type_id, company_id=self.company_id.id)
+            bom = boms.filtered(
+                lambda b: b.product_id.id == current_line.product_id.id or
+                          b.product_tmpl_id.id == current_line.product_id.product_tmpl_id.id)
```

Comment:
Calling `filtered` will loop over all records in `boms` each time.

It's much better if you produce a couple of dicts to access directly through product and template ids respectively, outside of the loop. Then here in the loop, directly assign the one via product id, and via template id if the 1st one failed.